src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,160 @@
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size()) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> oldNode = iterateToIndex(index);
+            newNode.next = oldNode;
+            newNode.prev = oldNode.prev;
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException(index + " out of bounds for " + size());
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node: list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = iterateToIndex(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = iterateToIndex(index);
+        T oldNode = node.value;
+        node.value = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removeValue;
+        if (index == 0) {
+            removeValue = head.value;
+            head = head.next;
+        } else if (index == size() - 1) {
+            removeValue = tail.value;
+            tail = tail.prev;
+        } else {
+            Node<T> oldNode = iterateToIndex(index);
+            removeValue = oldNode.value;
+            unlink(oldNode);
+        }
+        size--;
+        return removeValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        for (int i = 0; i < size(); i++) {
+            if (object == get(i) || object != null && object.equals(get(i))) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
+    private Node<T> iterateToIndex(int index) {
+        if (index <= size() / 2) {
```

Comment:
```suggestion
        if (index <= size / 2) {
```
remove all method calls to field addressing